ocaml/libs/tracing/tracing_export.ml

```diff
@@ -0,0 +1,346 @@
+
+      let compress_tracing_files = ref true
+
+      let set_trace_log_dir dir = trace_log_dir := dir
+
+      let get_trace_log_dir () = !trace_log_dir
+
+      let set_max_file_size size = max_file_size := size
+
+      let set_compress_tracing_files enabled = compress_tracing_files := enabled
+
+      let file_name = ref None
+
+      let lock = Mutex.create ()
+
+      let new_file_name () =
+        let date = Ptime_clock.now () |> Ptime.to_rfc3339 ~frac_s:6 in
+        let ( // ) = Filename.concat in
+        let name =
+          !trace_log_dir
+          // String.concat "-" [get_service_name (); !host_id; date]
+          ^ ".ndjson"
+        in
+        file_name := Some name ;
+        name
+
+      let with_fd file_name =
+        Xapi_stdext_unix.Unixext.with_file file_name
+          [O_WRONLY; O_CREAT; O_APPEND]
+          0o700
+
+      let write fd str =
+        let content = str ^ "\n" in
+        ignore @@ Unix.write_substring fd content 0 (String.length content)
+
+      let export json =
+        try
+          let file_name =
+            match !file_name with None -> new_file_name () | Some x -> x
+          in
+          Xapi_stdext_unix.Unixext.mkdir_rec (Filename.dirname file_name) 0o700 ;
+          let@ fd = file_name |> with_fd in
+          write fd json ;
+          if (Unix.fstat fd).st_size >= !max_file_size then (
+            debug "Tracing: Rotating file %s > %d" file_name !max_file_size ;
+            if !compress_tracing_files then
+              Zstd.Fast.compress_file Zstd.Fast.compress ~file_path:file_name
+                ~file_ext:"zst" ;
+            ignore @@ new_file_name ()
+          ) ;
+          Ok ()
+        with e -> Error e
+
+      let with_stream f =
+        Xapi_stdext_threads.Threadext.Mutex.execute lock (fun () -> f export)
+    end
+
+    module Http = struct
+      module Request = Cohttp.Request.Make (Cohttp_posix_io.Buffered_IO)
+      module Response = Cohttp.Response.Make (Cohttp_posix_io.Buffered_IO)
+
+      let export ~url json =
+        try
+          let body = json in
+          let headers =
+            Cohttp.Header.of_list
+              ([
+                 ("Content-Type", "application/json")
+               ; ("Content-Length", string_of_int (String.length body))
+               ]
+              @
+              match Uri.host url with
+              | None ->
+                  []
+              | Some h ->
+                  let port =
+                    match Uri.port url with
+                    | Some p ->
+                        ":" ^ string_of_int p
+                    | None ->
+                        ""
+                  in
+                  [("Host", h ^ port)]
+              )
+          in
+          Open_uri.with_open_uri url (fun fd ->
+              let request =
+                Cohttp.Request.make ~meth:`POST ~version:`HTTP_1_1 ~headers url
+              in
+              let ic = Unix.in_channel_of_descr fd in
```

Comment:
How and when are these closed? Add a comment.

Reply:
https://github.com/xapi-project/xen-api/pull/5551/commits/17fdf8bffb726ba3421a07f8fee35be06a69ad1b Added in here!